### ISSUE TITLE
bug(android): attempting to open a package from Samsung 'My Files' app results in an error 'You don't have any apps that can open this type of file'

### ISSUE THREAD

**Issue by mcdurdin** [keymanapp/keyman#10133]
### Describe the bug

![Screenshot_20231205_123603_My Files](https://github.com/keymanapp/keyman/assets/4498365/08bb5687-a6f1-4ac9-a38d-c524964e6965)

### Keyman apps

- [X] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.217-alpha-test-10103

**Comment by darcywong00** [comment]
Previous issue #629 about Samsung's "My Files" browser.

As a workaround, does the Chrome Downloads allow you to install the .kmp file?

**Comment by mcdurdin** [comment]
Trying to install from Chrome Downloads, I get a toast from Keyman: "storage request permission was denied. May fail to install key..." 

FWIW, any error which arises from a user action should be an alert rather than a toast.

**Comment by darcywong00** [comment]
Triaged with @mcdurdin to B17S1

**Comment by mcdurdin** [comment]
> Trying to install from Chrome Downloads, I get a toast from Keyman: "storage request permission was denied. May fail to install key..."

Screencap. And a small insight into my life if you look carefully.

https://github.com/keymanapp/keyman/assets/4498365/91526ff9-6aa9-42ae-8125-8134cd87810d

**Comment by darcywong00** [comment]
MyFiles creates the intent that Android Studio warns with
> [1235/FileExecute         ] openFile() ] There is no activity that can handle the intent. (type :  , data : content://media/external/file/11269)

Chrome Downloads does trigger the intent with
> START u0 {act=android.intent.action.VIEW dat=content://com.android.chrome.DownloadFileProvider/... typ=application/octet-stream flg=0x10000003 cmp=android/com.android.internal.app.ResolverAct

**Comment by darcywong00** [comment]
Retesting with 17.0.289 beta and I see this in the Logcat trying to launch a .kmp file from Samsung MyFiles:
```
com.sec.android.app.myfiles          I  [1351/FileExecute         ] 
    getOpenFileIntent() ] unsupported file type extension : kmp, mimeType : 
```

I'll call this a dupe of #629 which has some unpleasant caveats with the intent-filters

Per [SO reference](https://stackoverflow.com/a/77887488), Samsung MyFiles browser 

> The sad answer is that is no good solution but to use `*/*` and no file extension filtering.
>
>When Samsung's file manager tries opening a file whose mime type they don't recognize, they default to passing `""` (an empty string), which is impossible to filter for - setting an empty mime type in the intent filter creates a malformed APK that your device will refuse to install, so the only way to catch it is by using `*/*`.
>
>They probably should have defaulted to application/octet-stream, but that's not going to happen anytime soon - unless you work for Samsung, in which case, go change it :)

----

So, the only (uber unpleasant) alternative is to have an intent filter on **all** filetypes so the user could select Keyman for .kmp files.

Will close as won't fix - waiting on external

**Comment by darcywong00** [comment]
Re-opening to track for Future - still waiting on external Samsung update to MyFiles

**Comment by mcdurdin** [comment]
https://forum.developer.samsung.com/t/opening-files-with-unknown-content-type-from-my-files/30335

**Comment by mcdurdin** [comment]
Samsung have acknowledged that the issue is in their app but have declined to fix it. They don't consider Keyman a popular enough app to add specific support for it, and the hack that they've used for a long time is so embedded it can't really be fixed now, it seems.

I think we go ahead and close this as 'wont-fix' since it's outside our control and if necessary advise our users that they avoid the buggy My Files app, and instead browse for .kmp to install from within the Keyman app.

**Comment by mcdurdin** [comment]
Reopening and leaving forever in 'waiting-external' state because the issue is still present, even if we can't do anything about it.

**Comment by bamsbamx** [comment]
@mcdurdin 6th of June of 2025 and this issue still persists. Apparently the My Files app wont start any intent at all if it does not acknowledge the extension or mimetype

**Comment by mcdurdin** [comment]
Yes, .kmp now has a [mime type](https://www.iana.org/assignments/media-types/application/vnd.keyman.kmp+zip), but this will take a while to percolate through the ecosystems. We do plan to add this into Keyman for Android in the 19.0 release cycle (just created #14122 to track this, thanks for the reminder!)

**Comment by mcdurdin** [comment]
Another suggested mitigation is to use Files by Google app instead of Samsung Files -- this works fine. (As tested with #16392, Samsung Files still has the same issue)

**Comment by jahorton** [comment]
Associating our specific MIME type does not appear to mitigate this with Samsung's "My Files".  It's likely that we'd have to associate with literally any file per https://github.com/kiwix/kiwix-android/issues/4002 if we wanted to be able to install directly from Samsung's "My Files".

**Comment by mcdurdin** [comment]
And we don't want to associate with any file, because that's an unpleasant user experience too -- injecting our app where it is not wanted.

**Comment by mcdurdin** [comment]
See #10967 to follow the glacial deployment of mime types -- it looks like Android hasn't had a full media-types (aka mime types) database update since 2019... Given this, our existing mitigation will have to suffice for Samsung; adhoc install of .kmp files is not regular keyboard install pathway so damage is less significant.
